src/libraries/System.Linq/src/System/Linq/Cast.cs

```diff
@@ -63,12 +63,14 @@ private static IEnumerable<TResult> CastIterator<TResult>(IEnumerable source)
         }
 
         [DebuggerDisplay("Count = {Count}")]
-        private sealed partial class CastIListIterator<TResult>(IList source) : Iterator<TResult>, IList<TResult>, IReadOnlyList<TResult>
+        private sealed partial class CastICollectionIterator<TResult>(ICollection source) : Iterator<TResult>, IReadOnlyCollection<TResult>
```

Comment:
`IReadOnlyCollection<T>` isn't recognized by most Linq methods, including `TryGetNonEnumeratedCount`. Do we really need to add an implementation?